File: frontend/src/engine/capabilities/pageLayout.ts
/**
 * pageLayout – Configure page layout settings for printing.
 *
 * Office.js notes:
 * - WorksheetPageLayout API is in ExcelApi 1.9+.
 * - Margins are set in inches.
 * - Orientation and paper size are enums.
 * - setPrintArea() defines the range that will be printed.
 */

import { CapabilityMeta, PageLayoutParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "pageLayout",
  description: "Configure page layout — margins, orientation, paper size, print area, gridlines",
  mutates: false,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.9",
};

async function handler(
  context: Excel.RequestContext,
  params: PageLayoutParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { sheetName, margins, orientation, paperSize, printArea, showGridlines, printGridlines } = params;

  if (options.dryRun) {
    const parts: string[] = [];
    if (margins) parts.push("margins");
    if (orientation) parts.push(`orientation=${orientation}`);
    if (paperSize) parts.push(`paperSize=${paperSize}`);
    if (printArea) parts.push(`printArea=${printArea}`);
    if (showGridlines !== undefined) parts.push(`showGridlines=${showGridlines}`);
    if (printGridlines !== undefined) parts.push(`printGridlines=${printGridlines}`);
    return {
      stepId: "",
      status: "success",
      message: `Would set page layout: ${parts.join(", ")}`,
    };
  }

  options.onProgress?.("Configuring page layout...");

  // Resolve target sheet
  let sheet: Excel.Worksheet;
  if (sheetName) {
    const ws = context.workbook.worksheets.getItemOrNullObject(sheetName);
    ws.load("isNullObject");
    await context.sync();
    if (ws.isNullObject) {
      return {
        stepId: "",
        status: "error",
        message: `Sheet "${sheetName}" not found. Please check the sheet name.`,
      };
    }
    sheet = ws;
  } else {
    sheet = context.workbook.worksheets.getActiveWorksheet();
  }

  const layout = sheet.pageLayout;

  // Margins (in inches)
  if (margins) {
    if (margins.top !== undefined) layout.topMargin = margins.top * 72;
    if (margins.bottom !== undefined) layout.bottomMargin = margins.bottom * 72;
    if (margins.left !== undefined) layout.leftMargin = margins.left * 72;
    if (margins.right !== undefined) layout.rightMargin = margins.right * 72;
    if (margins.header !== undefined) layout.headerMargin = margins.header * 72;
    if (margins.footer !== undefined) layout.footerMargin = margins.footer * 72;
  }

  // Orientation
  if (orientation) {
    layout.orientation = orientation === "landscape"
      ? Excel.PageOrientation.landscape
      : Excel.PageOrientation.portrait;
  }

  // Paper size
  if (paperSize) {
    const sizeMap: Record<string, Excel.PaperType> = {
      letter: Excel.PaperType.letter,
      legal: Excel.PaperType.legal,
      a3: Excel.PaperType.a3,
      a4: Excel.PaperType.a4,
      a5: Excel.PaperType.a5,
      b4: Excel.PaperType.b4,
      b5: Excel.PaperType.b5,
      tabloid: Excel.PaperType.tabloid,
    };
    const paperType = sizeMap[paperSize.toLowerCase()];
    if (paperType !== undefined) {
      layout.paperSize = paperType;
    }
  }

  // Print area
  if (printArea) {
    layout.setPrintArea(printArea);
  }

  // Gridlines
  if (showGridlines !== undefined) {
    sheet.showGridlines = showGridlines;
  }
  if (printGridlines !== undefined) {
    layout.printGridlines = printGridlines;
  }

  await context.sync();

  const changes: string[] = [];
  if (margins) changes.push("margins");
  if (orientation) changes.push(orientation);
  if (paperSize) changes.push(paperSize);
  if (printArea) changes.push(`print area ${printArea}`);
  if (showGridlines !== undefined) changes.push(`gridlines ${showGridlines ? "on" : "off"}`);
  if (printGridlines !== undefined) changes.push(`print gridlines ${printGridlines ? "on" : "off"}`);

  return {
    stepId: "",
    status: "success",
    message: `Page layout updated: ${changes.join(", ")}${sheetName ? ` on "${sheetName}"` : ""}`,
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.9) ────────────────────────────────────
// WorksheetPageLayout is 1.9+. Below that, Office.js offers no programmatic
// handle on margins, orientation, paper size, or print area. `showGridlines`
// and `sheet.getRange(...).printArea` don't exist on 1.3 either. The correct
// behavior is a graceful skip — page layout is print-only and doesn't affect
// on-screen data; the plan should keep running. We emit a descriptive
// success with a warning so the audit trail shows what was skipped.
async function fallback(
  _context: Excel.RequestContext,
  params: PageLayoutParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: "Would skip page-layout (legacy fallback; requires ExcelApi 1.9+).",
    };
  }

  options.onProgress?.("Legacy-Excel mode: page layout unavailable, skipping...");

  const requested: string[] = [];
  if (params.margins) requested.push("margins");
  if (params.orientation) requested.push(`orientation=${params.orientation}`);
  if (params.paperSize) requested.push(`paperSize=${params.paperSize}`);
  if (params.printArea) requested.push(`printArea=${params.printArea}`);
  if (params.showGridlines !== undefined) requested.push(`showGridlines=${params.showGridlines}`);
  if (params.printGridlines !== undefined) requested.push(`printGridlines=${params.printGridlines}`);

  return {
    stepId: "",
    status: "success",
    message:
      `Page layout update skipped — WorksheetPageLayout requires ExcelApi 1.9+. ` +
      `Requested: ${requested.join(", ") || "(nothing)"}. Configure manually via File › Page Setup ` +
      `(legacy-Excel fallback).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/page_layout.py
"""pageLayout — configure sheet page layout (margins, orientation, paper size, print area, gridlines)."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry


# XlPageOrientation: xlPortrait=1, xlLandscape=2.
_ORIENTATION = {"portrait": 1, "landscape": 2}

# XlPaperSize common values.
_PAPER = {
    "letter": 1,     # xlPaperLetter
    "legal": 5,      # xlPaperLegal
    "a3": 8,         # xlPaperA3
    "a4": 9,         # xlPaperA4
    "a5": 11,        # xlPaperA5
    "b4": 12,        # xlPaperB4
    "b5": 13,        # xlPaperB5
    "tabloid": 3,    # xlPaperTabloid (11x17)
}


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    sheet_name = params.get("sheetName")
    margins = params.get("margins")
    orientation = params.get("orientation")
    paper_size = params.get("paperSize")
    print_area = params.get("printArea")
    show_gridlines = params.get("showGridlines")
    print_gridlines = params.get("printGridlines")

    book = ctx.workbook_handle

    if ctx.dry_run:
        parts: list[str] = []
        if margins:
            parts.append("margins")
        if orientation:
            parts.append(f"orientation={orientation}")
        if paper_size:
            parts.append(f"paperSize={paper_size}")
        if print_area:
            parts.append(f"printArea={print_area}")
        if show_gridlines is not None:
            parts.append(f"showGridlines={show_gridlines}")
        if print_gridlines is not None:
            parts.append(f"printGridlines={print_gridlines}")
        return {
            "status": "preview",
            "message": f"Would set page layout: {', '.join(parts) or 'no-op'}.",
        }

    # Resolve the target sheet.
    if sheet_name:
        try:
            sheet = book.sheets[sheet_name]
        except Exception:
            return {
                "status": "error",
                "message": f"Sheet {sheet_name!r} not found. Please check the sheet name.",
            }
    else:
        sheet = book.sheets.active

    try:
        page_setup = sheet.api.PageSetup

        # Margins — TS multiplies inches by 72 (points-per-inch). Match that.
        if margins:
            if margins.get("top") is not None:
                page_setup.TopMargin = margins["top"] * 72
            if margins.get("bottom") is not None:
                page_setup.BottomMargin = margins["bottom"] * 72
            if margins.get("left") is not None:
                page_setup.LeftMargin = margins["left"] * 72
            if margins.get("right") is not None:
                page_setup.RightMargin = margins["right"] * 72
            if margins.get("header") is not None:
                page_setup.HeaderMargin = margins["header"] * 72
            if margins.get("footer") is not None:
                page_setup.FooterMargin = margins["footer"] * 72

        # Orientation.
        if orientation:
            value = _ORIENTATION.get(orientation.lower(), _ORIENTATION["portrait"])
            page_setup.Orientation = value

        # Paper size.
        if paper_size:
            paper = _PAPER.get(paper_size.lower())
            if paper is not None:
                page_setup.PaperSize = paper

        # Print area.
        if print_area:
            page_setup.PrintArea = print_area

        # Gridlines — showGridlines affects the display (ActiveWindow), printGridlines
        # is a print-time page setup flag.
        if show_gridlines is not None:
            try:
                sheet.activate()
                book.app.api.ActiveWindow.DisplayGridlines = bool(show_gridlines)
            except Exception:
                pass
        if print_gridlines is not None:
            page_setup.PrintGridlines = bool(print_gridlines)

    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"Page layout failed: {exc}",
            "error": str(exc),
        }

    changes: list[str] = []
    if margins:
        changes.append("margins")
    if orientation:
        changes.append(orientation)
    if paper_size:
        changes.append(paper_size)
    if print_area:
        changes.append(f"print area {print_area}")
    if show_gridlines is not None:
        changes.append(f"gridlines {'on' if show_gridlines else 'off'}")
    if print_gridlines is not None:
        changes.append(f"print gridlines {'on' if print_gridlines else 'off'}")

    suffix = f' on "{sheet_name}"' if sheet_name else ""
    return {
        "status": "success",
        "message": f"Page layout updated: {', '.join(changes) or 'no-op'}{suffix}.",
        "outputs": {"sheet": sheet.name},
    }


registry.register("pageLayout", handler, mutates=False, affects_formatting=True)
